pkg/querysharding/analyzer.go

```diff
@@ -6,21 +6,65 @@ package querysharding
 import (
 	"fmt"
 
+	lru "github.com/hashicorp/golang-lru"
 	"github.com/prometheus/prometheus/promql/parser"
 )
 
 // QueryAnalyzer is an analyzer which determines
 // whether a PromQL Query is shardable and using which labels.
+
+type Analyze interface {
+	Analyze(string) (QueryAnalysis, error)
+}
+
 type QueryAnalyzer struct{}
 
+type CachedQueryAnalyzer struct {
+	analyzer *QueryAnalyzer
+	cache    *lru.Cache
+}
+
 var nonShardableFuncs = []string{
 	"label_join",
 	"label_replace",
 }
 
 // NewQueryAnalyzer creates a new QueryAnalyzer.
-func NewQueryAnalyzer() *QueryAnalyzer {
-	return &QueryAnalyzer{}
+func NewQueryAnalyzer() (Analyze, error) {
+	cache, err := lru.New(256)
+
+	if err != nil {
+		return &QueryAnalyzer{}, err
+	}
+
+	return &CachedQueryAnalyzer{
+		analyzer: &QueryAnalyzer{},
+		cache:    cache,
+	}, nil
+}
+
+type cachedValue struct {
+	QueryAnalysis QueryAnalysis
+	err           error
+}
+
+func (a *CachedQueryAnalyzer) Analyze(query string) (QueryAnalysis, error) {
+	if a.cache.Contains(query) {
+		value, ok := a.cache.Get(query)
+		if !ok {
+			return QueryAnalysis{}, fmt.Errorf("failed to fetch query:%s from cache", query)
```

Comment:
If it is not okay, this means that the result is not cached? We should let the analyzer analyzes the query right

Reply:
I think, if it is not okay that means there is some issue is fetching the value from cache, and `Contains` function ensures that the query is already present inside the cache after which `Get` function starts its work.

Reply:
I see, but even if it is not cached, let's not return error.
We should let analyzer analyze again instead of failing the whole query.

Reply:
Understood